File: backend/src/test/java/ca/metricalsky/winston/repository/ChannelRepositoryTest.java
package ca.metricalsky.winston.repository;

import ca.metricalsky.winston.database.repository.channel.ChannelRepository;
import ca.metricalsky.winston.test.annotations.RepositoryTest;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.jdbc.Sql;

import static org.assertj.core.api.Assertions.assertThat;

@RepositoryTest
@Sql("classpath:sql/channels.sql")
class ChannelRepositoryTest {

    private static final String CHANNEL_ID = "channelId";
    private static final String CUSTOM_URL = "@customUrl";
    private static final String CUSTOM_URL_NOT_FOUND = "@notFound";

    @Autowired
    private ChannelRepository channelRepository;

    @Test
    void findIdByCustomUrl() {
        var result = channelRepository.findIdByCustomUrl(CUSTOM_URL);

        assertThat(result)
                .hasValue(CHANNEL_ID);
    }

    @Test
    void findIdByCustomUrl_notFound() {
        var result = channelRepository.findIdByCustomUrl(CUSTOM_URL_NOT_FOUND);

        assertThat(result)
                .isEmpty();
    }
}

File: database/src/test/kotlin/ca/metricalsky/winston/database/repository/channel/ChannelRepositoryTest.kt
package ca.metricalsky.winston.database.repository.channel

import ca.metricalsky.winston.database.entity.channel.ChannelEntity
import ca.metricalsky.winston.database.test.annotation.DatabaseTest
import ca.metricalsky.winston.database.test.faker.DatabaseFaker
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager

@DatabaseTest
class ChannelRepositoryTest(
    channelRepository: ChannelRepository,
    entityManager: TestEntityManager,
): WordSpec({

    val faker = DatabaseFaker()
    var channel: ChannelEntity? = null

    beforeEach {
        channel = entityManager.persist(faker.channel().completeEntity())
    }

    "findIdByCustomUrl()" should {

        "return a channel when given its custom URL" {
            val result = channelRepository.findIdByCustomUrl(channel?.customUrl!!)
            result shouldBe channel.id
        }

        "return null when given an invalid custom URL" {
            val result = channelRepository.findIdByCustomUrl("@notFound")
            result shouldBe null
        }
    }
})
